gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/DagProcessingEngine.java

```diff
@@ -0,0 +1,128 @@
+import org.apache.gobblin.runtime.api.DagActionStore;
+import org.apache.gobblin.runtime.api.MultiActiveLeaseArbiter;
+import org.apache.gobblin.service.ServiceConfigKeys;
+import org.apache.gobblin.service.modules.flowgraph.Dag;
+import org.apache.gobblin.service.modules.orchestration.proc.DagProc;
+import org.apache.gobblin.service.modules.orchestration.task.DagTask;
+import org.apache.gobblin.service.modules.spec.JobExecutionPlan;
+import org.apache.gobblin.util.ConfigUtils;
+
+
+/**
+ * Responsible for polling {@link DagTask}s from {@link DagTaskStream} and processing the {@link org.apache.gobblin.service.modules.flowgraph.Dag}
+ * based on the type of {@link DagTask} which is determined by the {@link org.apache.gobblin.runtime.api.DagActionStore.DagAction}.
+ * Each {@link DagTask} acquires a lease for the {@link org.apache.gobblin.runtime.api.DagActionStore.DagAction}.
+ * The {@link DagProcFactory} then provides the appropriate {@link DagProc} associated with the {@link DagTask}.
+ * The actual work or processing is done by the {@link DagProc#process(DagManagementStateStore, int, long)}
+ */
+
+@Alpha
+@Slf4j
+public class DagProcessingEngine {
+  public static final String GOBBLIN_SERVICE_DAG_PROCESSING_ENGINE_PREFIX = ServiceConfigKeys.GOBBLIN_SERVICE_PREFIX + "dagProcessingEngine.";
+  public static final String NUM_THREADS_KEY = GOBBLIN_SERVICE_DAG_PROCESSING_ENGINE_PREFIX + "numThreads";
+  public static final String JOB_STATUS_POLLING_INTERVAL_KEY = GOBBLIN_SERVICE_DAG_PROCESSING_ENGINE_PREFIX + "pollingInterval";
+  private static final Integer DEFAULT_NUM_THREADS = 3;
+  private static final Integer DEFAULT_JOB_STATUS_POLLING_INTERVAL = 10;
+
+  private final DagTaskStream dagTaskStream;
+  Optional<EventSubmitter> eventSubmitter;
+
+  public DagProcessingEngine(Config config, DagTaskStream dagTaskStream, DagProcFactory dagProcFactory, MultiActiveLeaseArbiter multiActiveLeaseArbiter) {
+    MetricContext metricContext = Instrumented.getMetricContext(ConfigUtils.configToState(ConfigFactory.empty()), getClass());
+    this.eventSubmitter = Optional.of(new EventSubmitter.Builder(metricContext, "org.apache.gobblin.service").build());
+    Integer numThreads = ConfigUtils.getInt(config, NUM_THREADS_KEY, DEFAULT_NUM_THREADS);
+    ScheduledExecutorService scheduledExecutorPool = Executors.newScheduledThreadPool(numThreads);
+    Integer pollingInterval = ConfigUtils.getInt(config, JOB_STATUS_POLLING_INTERVAL_KEY, DEFAULT_JOB_STATUS_POLLING_INTERVAL);
+    this.dagTaskStream = dagTaskStream;
+
+    for (int i=0; i < numThreads; i++) {
+      DagProcEngineThread dagProcEngineThread = new DagProcEngineThread(dagTaskStream, dagProcFactory);
+      scheduledExecutorPool.scheduleAtFixedRate(dagProcEngineThread, 0, pollingInterval, TimeUnit.SECONDS);
+    }
+  }
+
+  public void addNewDag(Dag<JobExecutionPlan> dag) {
+    this.dagTaskStream.addDagAction(DagManagerUtils.createDagAction(dag, DagActionStore.FlowActionType.LAUNCH));
```

Comment:
the DPE should be pulling `DagTask`s from the stream (using `next()`, since it's an `Iterator`).  AFAICT, is should be making no other calls to the DTS